File: segmentSupport/src/main/java/com/clumob/segment/support/pager/viewpager/SegmentStatePagerAdapter.java
package com.clumob.segment.support.pager.viewpager;

import android.content.res.Configuration;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.Nullable;

import com.clumob.listitem.controller.source.ItemController;
import com.clumob.listitem.controller.source.ItemControllerSource;
import com.clumob.listitem.controller.source.SourceUpdateEvent;
import com.clumob.segment.controller.SegmentInfo;
import com.clumob.segment.manager.Segment;
import com.clumob.segment.support.pager.SegmentItemProvider;

import java.util.Deque;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Set;

import io.reactivex.observers.DisposableObserver;

import static com.clumob.segment.support.pager.viewpager.SegmentStatePagerAdapter.ItemSegmentPair.pair;

/**
 * Created by prashant.rathore on 02/07/18.
 */

public class SegmentStatePagerAdapter<T extends ItemController> extends SegmentPagerAdapter {

    private final ItemControllerSource<T> dataSource;
    private final SegmentItemProvider factory;
    private Set<ItemSegmentPair<T>> attachedSegments = new HashSet<>();
    private Handler mHandler = new Handler(Looper.getMainLooper());

    public SegmentStatePagerAdapter(ItemControllerSource<T> dataSource,
                                    SegmentItemProvider factory) {
        this.dataSource = dataSource;
        this.dataSource.setViewInteractor(createViewInteractor());
        this.factory = factory;
        this.dataSource.observeAdapterUpdates().subscribe(new DisposableObserver<SourceUpdateEvent>() {
            @Override
            public void onNext(SourceUpdateEvent sourceUpdateEvent) {
                if (sourceUpdateEvent.getType() == SourceUpdateEvent.Type.UPDATE_ENDS)
                    notifyDataSetChanged();
            }

            @Override
            public void onError(Throwable e) {
                e.printStackTrace();
            }

            @Override
            public void onComplete() {

            }
        });
    }

    private ItemControllerSource.ViewInteractor createViewInteractor() {
        return new ItemControllerSource.ViewInteractor() {
            Deque<Runnable> deque = new LinkedList();
            private boolean processingInProgress;

            public void processWhenSafe(Runnable runnable) {
                this.deque.add(runnable);
                if (!this.processingInProgress) {
                    this.processingInProgress = true;
                    this.processWhenQueueIdle();
                }

            }

            private void processWhenQueueIdle() {
                mHandler.post(new Runnable() {
                    public void run() {
                        if (isComputingLayout()) {
                            mHandler.post(this);
                        } else if (deque.peekFirst() != null) {
                            Runnable runnable = (Runnable) deque.pollFirst();
                            runnable.run();
                            mHandler.post(this);
                        } else {
                            processingInProgress = false;
                        }

                    }
                });
            }

            public void cancelOldProcess(Runnable runnable) {
                mHandler.removeCallbacks(runnable);
            }
        };
    }

    private boolean isComputingLayout() {
        return false;
    }

    @Override
    public Object instantiateItem(int index) {
        T item = getItem(index);
        Segment<?, ?> segment = factory.provide(item);
        ItemSegmentPair pair = pair(segment, item);
        attachedSegments.add(pair);
        segment.onCreate();
        dataSource.onItemAttached(index);
        pair.itemController.onAttach(pair.segment);
        return pair;
    }

    @Override
    public int getCount() {
        return dataSource.getItemCount();
    }

    public Set<ItemSegmentPair<T>> getAttachedSegments() {
        return attachedSegments;
    }

    @Override
    protected Segment<?, ?> retrieveSegmentFromObject(Object object) {
        return ((ItemSegmentPair) object).segment;
    }

    // ToDO: The lookup algorightm is slow;
    @Override
    public int computeItemPosition(Object inputItem) {
        final long id = ((ItemSegmentPair) inputItem).itemController.getId();
        int itemCount = dataSource.getItemCount();
        for (int i = 0; i < itemCount; i++) {
            T item = getItem(i);
            if (item.getId() == id) {
                return i;
            }
        }
        return POSITION_NONE;
    }


    @Override
    public void destroyItem(Object object) {
        super.destroyItem(object);
        ItemSegmentPair pair = (ItemSegmentPair) object;
        pair.itemController.onDetach(pair.segment);
        pair.segment.onDestroy();
        attachedSegments.remove(pair);
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstance) {
        super.onCreate(savedInstance);
        for (ItemSegmentPair<T> segment : attachedSegments) {
            segment.segment.onCreate();
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        for (ItemSegmentPair<T> segment : attachedSegments) {
            segment.segment.onStart();
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        for (ItemSegmentPair<T> segment : attachedSegments) {
            segment.segment.onPause();
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outBundle) {
//        for(Segment segment : attachedSegments) {
//        }
    }

    @Override
    public void onConfigurationChanged(Configuration configuration) {
        for (ItemSegmentPair<T> segment : attachedSegments) {
            segment.segment.onConfigurationChanged(configuration);
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        for (ItemSegmentPair<T> segment : attachedSegments) {
            segment.segment.onStop();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        for (ItemSegmentPair<T> segment : attachedSegments) {
            segment.segment.onDestroy();
        }
    }

    public T getItem(int position) {
        return dataSource.getItem(position);
    }

    public ItemControllerSource<T> getDataSource() {
        return dataSource;
    }

    static class ItemSegmentPair<T extends ItemController> {

        final Segment<?, ?> segment;
        final T itemController;

        public ItemSegmentPair(Segment<?, ?> segment, T itemController) {
            this.segment = segment;
            this.itemController = itemController;
        }

        static <T extends ItemController> ItemSegmentPair<T> pair(Segment<?, ?> segment, T itemController) {
            return new ItemSegmentPair<T>(segment, itemController);
        }
    }
}

File: segmentSupport/src/main/java/com/clumob/segment/support/pager/viewpager/SegmentStatePagerAdapter.kt
package com.clumob.segment.support.pager.viewpager

import android.content.res.Configuration
import android.os.Handler
import android.os.Looper
import androidx.viewpager.widget.PagerAdapter
import com.clumob.listitem.controller.source.ItemController
import com.clumob.listitem.controller.source.ItemControllerSource
import com.clumob.listitem.controller.source.SourceUpdateEvent
import com.clumob.segment.controller.list.SegmentItemController
import com.clumob.segment.manager.Segment
import com.clumob.segment.support.pager.SegmentItemProvider
import io.reactivex.observers.DisposableObserver
import java.util.*

open class SegmentStatePagerAdapter<T : SegmentItemController>(val dataSource : ItemControllerSource<T>,
                                    val factory: SegmentItemProvider) : SegmentPagerAdapter() {

    private val attachedSegments: MutableSet<ItemSegmentPair> = HashSet()
    private val mHandler = Handler(Looper.getMainLooper())

    init {
        dataSource.viewInteractor = (createViewInteractor())
        dataSource.observeAdapterUpdates().subscribe(object : DisposableObserver<SourceUpdateEvent?>() {
            override fun onNext(sourceUpdateEvent: SourceUpdateEvent) {
                if (sourceUpdateEvent.type == SourceUpdateEvent.Type.UPDATE_ENDS) notifyDataSetChanged()
            }

            override fun onError(e: Throwable) {
                e.printStackTrace()
            }

            override fun onComplete() {}
        })
    }

    private fun createViewInteractor() : ItemControllerSource.ViewInteractor {
        return object : ItemControllerSource.ViewInteractor {
            var deque: Deque<Runnable> = LinkedList()
            private var processingInProgress = false
            override fun processWhenSafe(runnable: Runnable) {
                deque.add(runnable)
                if (!processingInProgress) {
                    processingInProgress = true
                    this.processWhenQueueIdle()
                }
            }

            private fun processWhenQueueIdle() {
                mHandler.post(object : Runnable {
                    override fun run() {
                        if (isComputingLayout()) {
                            mHandler.post(this);
                        } else if (deque.peekFirst() != null) {
                            val runnable = deque.pollFirst();
                            runnable.run();
                            mHandler.post(this);
                        } else {
                            processingInProgress = false;
                        }
                    }
                });
            }

            override fun cancelOldProcess(runnable: Runnable) {
                mHandler.removeCallbacks(runnable)
            }
        };
    }

    private fun isComputingLayout(): Boolean {
        return false
    }

    class ItemSegmentPair(val segment: Segment, val itemController: ItemController?) {

        companion object {
            fun pair(segment: Segment, itemController: ItemController?): ItemSegmentPair {
                return ItemSegmentPair(segment, itemController)
            }
        }

    }

    override fun instantiateItem(index: Int): Any {
        val item = getItem(index)
        val segment = factory.provide(item)
        val pair = ItemSegmentPair.pair(segment, item)
        attachedSegments.add(pair)
        segment.onCreate()
        dataSource.onItemAttached(index)
        pair.itemController!!.onAttach(pair.segment)
        return pair
    }

    override fun getCount(): Int {
        return dataSource.itemCount
    }

    fun getAttachedSegments(): Set<ItemSegmentPair?>? {
        return attachedSegments
    }

    override fun retrieveSegmentFromObject(item: Any): Segment {
        return (item as ItemSegmentPair).segment
    }

    // ToDO: The lookup algorightm is slow;
    override fun computeItemPosition(inputItem: Any): Int {
        val id = (inputItem as ItemSegmentPair).itemController!!.id
        val itemCount = dataSource.itemCount
        for (i in 0 until itemCount) {
            val item = getItem(i)
            if (item.id == id) {
                return i
            }
        }
        return PagerAdapter.POSITION_NONE
    }


    override fun destroyItem(item: Any) {
        super.destroyItem(item)
        val pair = item as ItemSegmentPair
        pair.itemController!!.onDetach(pair.segment)
        pair.segment.onDestroy()
        attachedSegments.remove(pair)
    }


    override fun onCreate() {
        super.onCreate()
        for (segment in attachedSegments) {
            segment.segment.onCreate()
        }
    }

    override fun onStart() {
        super.onStart()
        for (segment in attachedSegments) {
            segment.segment.onStart()
        }
    }

    override fun onPause() {
        super.onPause()
        for (segment in attachedSegments) {
            segment.segment.onPause()
        }
    }

    override fun onConfigurationChanged(configuration: Configuration?) {
        for (segment in attachedSegments) {
            segment.segment.onConfigurationChanged(configuration)
        }
    }

    override fun onStop() {
        super.onStop()
        for (segment in attachedSegments) {
            segment.segment.onStop()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        for (segment in attachedSegments) {
            segment.segment.onDestroy()
        }
    }

    fun getItem(position: Int): T {
        return dataSource.getItem(position)
    }

}
